text/0000-fragment-specifiers-for-generic-arguments.md

```diff
@@ -0,0 +1,155 @@
+- Feature Name: fragment-specifiers-for-generic-arguments
+- Start Date: 2023-05-31
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Right now, there is no support for parsing the syntax of generic
+parameters/arguments in declarative macros. This makes it difficult to
+impossible to write a declarative macro that handles arbitrary generics easily.
+I propose adding fragment specifiers to parse a generic parameter definition as
+a whole and also parts of the definition.
+
+# Motivation
+[motivation]: #motivation
+
+I personally encountered this issue when attempting to write a declarative macro
+to implement a trait on a type. I wanted to write something like this minimal
+toy example:
+
+```rust
+macro_rules! implement_debug {
+    { $params:generic $type:ty } => {
+        impl $params Debug for $ty {
+            /* .. implementation goes here .. */
+        }
+    };
+}
+
+struct Container<'a, T>(&'a T);
+
+implement_debug!(<'a, T: Debug + 'a> Container<'a, T>);
+```
+
+However, with the current state of declarative macros, there's no way to parse
+arbitrary generic parameters in the body of the macro, forcing this macro of
+mine to be a procedural macro. However, declarative macros are easier to read
+and write, and this could be a declarative macro if there was only a way to
+parse generic parameters.
+
+Additionally, more complicated macros want to be able to parse each parameters
+and its bounds for use in various places, so I'd like this as well:
+
+```rust
+macro_rules! implement_debug {
+    { < $( $param:generic_param $( : $bound:generic_bound )? ),+ > $type:ty } => {
+        impl < $( $param $( : $bound )? ),+ > Debug for $ty {
+            /* .. implementation goes here .. */
+        }
+    };
+}
+
+struct Container<'a, T>(&'a T);
+
+implement_debug!(<'a, T: Debug + 'a> Container<'a, T>);
+```
+
+Any toy example will be obviously redundant with `:generic`, but more involved
+macros sometimes want it.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When explaining fragment specifiers, this can be explained by adding the new
+fragment specifiers and their descriptions:
+
+* `:generic`: A full set of generic parameters and their bounds (e.g. `<'a, T:
+  'a + SomeTrait, const N: usize>`)
+* `:generic_param`: A generic parameter (e.g. `'a`, `T`, or `const N`)
```

Comment:
`const N` does not work for the macro example above, `$param:generic_param $( : $bound:generic_bound )?`. Const generics have the syntax `const $param:ident : $ty:ty $(= $default:expr)?`. Note that it is a type instead of generic bounds after the semicolon, so syntactically it should also accept non-trait-like types like `const N: (i32, i32)`, though it's forbidden currently.

Reply:
This was a deliberate deviation from the default syntax to make it easier to parse an arbitrary set of generic parameters. If we use the fragment specifiers as I defined them in the example macro defined in the RFC (`< $( $param:generic_param $( : $bound:generic_bound )? ),+ > `, then this successfully matches `<const N: usize>` (`$param` being `const N` and `$bound` being `usize`.

If it matches as you're requesting that it match, then the same pattern will match this trait, but it will also allow in nonsense like `<const N: usize: 'a + SomeTrait>` which my approach avoids. I'm not sure how one could change the pattern to avoid allowing for nonsense like that.

Though also, my approach would match `<const N>`, which is also nonsense, so idk which one would be better to reject while matching the macro. I'm open to changing it if you have arguments for your way of splitting the definition between the two fragment specifiers being better.

To your last point, the definition that I give for `:generic_bounds` does explicitly allow for a type, so that syntax would be accepted.

Reply:
Your approach also matches `const N: 'a` and `'a: usize`